web/app.js

```diff
@@ -37,4 +37,122 @@ const NewFormButton = {
   },
 };
 
+const DragAndDrop = {
+  PLACEHOLDER_QUERY: '.drop-overlay',
+  FILE_FORM_QUERY: '.code-form--file',
+
+  init() {
+    this.lastTarget = null;
+    this.placeholder = document.querySelector(this.PLACEHOLDER_QUERY);
+
+    this.handleDragEnter();
+    this.handleDragLeave();
+    this.handleDragOver();
+    this.handleDrop();
+  },
+
+  handleDragEnter() {
+    window.addEventListener("dragenter", (event) => {
+      if (this.isFile(event)) {
+        this.lastTarget = event.target;
+        this.showPlaceholder();
+      }
+    });
+  },
+
+  handleDragLeave() {
+    window.addEventListener("dragleave", (event) => {
+      event.preventDefault();
+
+      if (event.target === this.lastTarget) {
+        this.hidePlaceholder();
+      }
+    });
+  },
+
+  handleDragOver() {
+    window.addEventListener("dragover", (event) => {
+      event.preventDefault();
+    });
+  },
+
+  handleDrop() {
+    window.addEventListener("drop", (event) => {
+      event.preventDefault();
+      this.hidePlaceholder();
+
+      const files = [...event.dataTransfer.files]
+        .filter(file => this.isReadable(file));
+
+      const fieldsetsToCreateNumber = files.length - this.getEmptyFieldsets().length;
+      this.createNewFieldsets(fieldsetsToCreateNumber);
+
+      const emptyFieldsets = this.getEmptyFieldsets();
+      emptyFieldsets.forEach((fieldset, index) => this.fillFieldset(fieldset, files[index]))
+    });
+  },
+
+  createNewFieldsets(number) {
+    for(let i = 0; i < number; i++) {
+      NewFormButton.onClick();
+    }
+  },
+
+  fillFieldset(fieldset, file) {
+    const input = fieldset.querySelector('input');
+    const textarea = fieldset.querySelector('textarea');
+
+    input.value = this.getFileName(file);
+
+    this.getFileContent(file).then(content => {
+      textarea.value = content;
+    });
+  },
+
+  getEmptyFieldsets() {
+    const fieldsets = [...document.querySelectorAll(this.FILE_FORM_QUERY)];
+
+    return fieldsets.filter(fieldset => this.isFieldsetEmpty(fieldset));
+  },
+
+  isFieldsetEmpty(fieldset) {
+    const inputValue = fieldset.querySelector('input').value;
+    const textareaValue = fieldset.querySelector('textarea').value;
+
+    return inputValue + textareaValue === '';
+  },
+
+  isFile(event) {
+    return event.dataTransfer.types.some(type => type === 'Files');
+  },
+
+  isReadable(file) {
+    return file.type.indexOf('text') == 0 || file.type.indexOf('application') == 0;
```

Comment:
There is many edge-cases that this does not cover and this also does not allow some files to be dropped. I think that we should create a whitelist of MIME types that will be allowed to be dropped. I can maintain that list, but you need to provide me the basis to work with.